CSI: enhance driver tests to update the drivers.

jsafrane opened the pull request:
We want to check that kubelet correctly handles driver update on a node and does not screw up sockets.

I extended an existing tests, because testing of update starts exactly the same as simply testing the driver and I want to avoid two tests where one is subset of the other one.

**Which issue(s) this PR fixes**
Fixes #70293

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
NONE
```

/sig storage
cc: @msau42 @pohly @davidz627

jsafrane commented:
/kind cleanup
/priority important-soon

jsafrane commented:
cc @bertinatto

davidz627 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -331,6 +364,21 @@ func startPausePod(cs clientset.Interface, t testsuites.StorageClassTest, ns str
+// getDriverDaemonSet finds *any* DaemonSet in framework's namespace and returns it.
+// It must be called just after driver installation, where the only DaemonSet in the
+// namespace must be the driver.
+func getDriverDaemonSet(f *framework.Framework) (*appsv1.DaemonSet, error) {
```
this function makes some significant assumptions:
1) Driver is deployed with Daemonset
2) Driver is deployed with *strictly* 1 Daemonset
3) No other non-driver Daemonsets deployed

I am not sure these are reasonable assumptions to make for arbitrary drivers

jsafrane replied:
Should it work for arbitrary drivers? I expect it to be used only in kubernetes e2e to test the Kubernetes part. I expect that each driver will have its own assumption how and when it can be updated.

If some generic mechanism is required (is it?), I can implement new `TestDriver` method to update a driver.

davidz627 replied:
maybe it doesn't have to but its worth calling out in a comment, any function written might get "resused" in some other code without true understanding of what it does just based on the function name.

davidz627 replied:
This part is failing on our external GCE PD tests as we bring up one driver in the default namespace as opposed to one in every test namespace. I wonder if maybe we can get the daemonset from a config specified namespace instead of using the test namespace. What do you think?

Also this test should be marked `[Serial] [Disruptive` if we make that change since it may affect usage of a shared driver.

/cc @msau42

msau42 replied:
Yes those sound like good enhancements

davidz627 replied:
https://github.com/kubernetes/kubernetes/pull/71527

davidz627 replied:
Did not mark as serial or disruptive because the k8s tests are not disruptive. And I don't think the external tests respect the "Disruptive" flag anyway (?). We should make sure the external CI is running tests serially anyway because of this test

davidz627 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -139,7 +141,12 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
 			})
 
-			It("should provision storage", func() {
+			It("should provision storage and allow driver update", func() {
```
i understand the want that one test is not strict subset of another; however, by doing this we have lost granularity of signals. When this test fails we don't have a simple signal whether the driver is failing to work in general or if it is failing to update.

Though I could be convinced this is ok because I guess you can look at error logs to see which error it really is. 
@msau42 @jsafrane what do you think?

msau42 replied:
would prefer two different test cases.  One for basic functionality, one for testing update

jsafrane replied:
Ack, looking at the test structure now, I need to create a separate test anyway.

davidz627 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -151,6 +158,32 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
+				framework.ExpectNoError(err, "Failed to update DaemonSet: %v", err)
+
+				By("Waiting for the update to complete")
+				selector := labels.SelectorFromSet(labels.Set(map[string]string{f.UniqueName: ""}))
```
not sure what this selector is for, seems to be the nil selector. It is not clear to me how this is getting the Daemonset pods.

wongma7 replied:
see some lines above https://github.com/kubernetes/kubernetes/pull/70578/files/c0770267040509587d7ba503eee2d6e55cce95cf#diff-0d9ecaa3e6a0297186ad33f57aad472eR176

jsafrane replied:
Yeah, adding unique label a few lines above triggers rolling update of the DaemonSet, which simulates driver update. Waiting for pods with the label then basically waits until the update is done and new pod(s) are running.

jsafrane commented:
I rewrote the test so it's separate from standard volume tests.

I had to change the HostPath driver yaml files to make dynamically provisioned volumes persistent after driver restart - they're now created in the host's `/tmp` instead of the driver's `/tmp`. @pohly, do you remember where the yaml files come from? I did not find them in kubernetes-csi/drivers

jsafrane committed:
CSI: enhance driver tests to update the drivers.

We want to check that kubelet correctly handles driver update
on a node.

jsafrane committed:
Make HostPath CSI driver persistent across restart.

The driver dynamically provisions its volumes in /tmp. To preserve the data
across driver restarts, the directory must be mapped to more persistent
place, /tmp on the host seems to be the safest choice.

jsafrane commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -332,7 +346,97 @@ func startPausePod(cs clientset.Interface, t testsuites.StorageClassTest, ns str
+		if pod == nil {
+			return
+		}
+		err := framework.WaitForPodNameRunningInNamespace(cs, pod.Name, pod.Namespace)
```
Added read + write.

jsafrane commented:
added test for #71424

AishSundar commented:
@jsafrane @msau42 are we attempting to get this test int 1.13?

msau42 commented:
I think we can get this in 1.13. The new tests are failing because of #71440

msau42 commented:
/lgtm

AishSundar commented:
/milestone v1.13

AishSundar commented:
/kind failing-test
/priority critical-urgent

Updating tags to get this into 1.13

saad-ali commented:
Thanks for getting this in @jsafrane.

/approve

Once the fix for https://github.com/kubernetes/kubernetes/issues/71434 is merged, rerunning the tests here should cause them to pass and for these tests to be merged.

saad-ali commented:
@jsafrane can your run this test in a tight loop with https://github.com/kubernetes/kubernetes/pull/71440 to make sure it is flake free?

I want to make sure this test doesn't introduce any more random flakiness while adding additional coverage.

saad-ali commented:
/hold

@jsafrane please remove hold once you've verified.

saad-ali commented:
Rerunning tests now that https://github.com/kubernetes/kubernetes/pull/71440 has merged:


/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-kops-aws

msau42 commented:
/retest

jsafrane commented:
/hold cancel

Run for 2 hours in 20 parallel threads, ~500 times in total. No flake.
